File: casl2/src/interpreter/interpreter.ts
import { advancePR, setPR } from "../casl2/procedures/registerAccessor"
import { Memory } from "../infra/memory"
import { END_ADDRESS, FlagRegister, GeneralRegister } from "../infra/register"

// 1 word
const NOP  = 0x00
const LD   = 0x14
const ADDA = 0x24
const SUBA = 0x25
const ADDL = 0x26
const SUBL = 0x27
const AND  = 0x34
const OR   = 0x35
const XOR  = 0x36
const CPA  = 0x44
const CPL  = 0x45
const POP  = 0x71
const RET  = 0x81

// 2 words
const LD2   = 0x10
const ST    = 0x11
const LAD   = 0x12
const ADDA2 = 0x20
const SUBA2 = 0x21
const ADDL2 = 0x22
const SUBL2 = 0x23
const AND2  = 0x30
const OR2   = 0x31
const XOR2  = 0x32
const CPA2  = 0x40
const CPL2  = 0x41
const SLA   = 0x50
const SRA   = 0x51
const SLL   = 0x52
const SRL   = 0x53
const JMI   = 0x61
const JNZ   = 0x62
const JZE   = 0x63
const JUMP  = 0x64
const JPL   = 0x65
const JOV   = 0x66
const PUSH  = 0x70
const CALL  = 0x80

// no support
// const SVC   = 0xF0

export class Interpreter {
  grMap: Map<string, GeneralRegister>
  FR: FlagRegister
  PR: GeneralRegister
  SP: GeneralRegister
  memory: Memory

  constructor(
    grMap: Map<string, GeneralRegister>,
    FR: FlagRegister,
    PR: GeneralRegister,
    SP: GeneralRegister,
    memory: Memory,
    code: ArrayBuffer,
  ) {
    this.grMap = grMap
    this.FR = FR
    this.PR = PR
    this.SP = SP
    this.memory = memory

    const dataView = new DataView(code)
    const startByteOffset = 2*this.PR.lookupLogical()
    for (let i = 0; i < dataView.byteLength; i++) {
      this.memory.storeLogical(startByteOffset + i, dataView.getUint8(i))
    }
  }

  private divide(byte: number): [number, number] {
      return [
        byte >> 4,
        byte & 0b1111,
      ]
  }

  private gr(x: number): number {
    return this.grMap.get(`GR${x}`)?.lookup() || 0
  }

  private grLogical(x: number): number {
    return this.grMap.get(`GR${x}`)?.lookupLogical() || 0
  }

  private storeGr(x: number, value: number) {
    this.grMap.get(`GR${x}`)?.store(value)
  }

  private storeGrLogical(x: number, value: number) {
    this.grMap.get(`GR${x}`)?.storeLogical(value)
  }

  private readWord(): [number, number] {
    const upperByte = this.memory.lookupLogical(2*this.PR.lookupLogical())
    const lowerByte = this.memory.lookupLogical(1 + 2*this.PR.lookupLogical())
    return [upperByte, lowerByte]
  }

  run() {
    let next = true
    while (next) {
      next = this.step()
    }
  }

  step(): boolean {
    const [opcode, operands] = this.readWord()
    advancePR(this.PR, 1)
    switch (opcode) {
      case NOP:
        break
      case LD:
        this.ld(operands)
        break
      case ADDA:
        this.adda(operands)
        break
      case SUBA:
        this.suba(operands)
        break
      case ADDL:
        this.addl(operands)
        break
      case SUBL:
        this.subl(operands)
        break
      case AND:
        this.and(operands)
        break
      case OR:
        this.or(operands)
        break
      case XOR:
        this.xor(operands)
        break
      case CPA:
        this.cpa(operands)
        break
      case CPL:
        this.cpl(operands)
        break
      case POP:
        this.pop(operands)
        break
      case RET:
        this.ret()
        break
      default:
        const [upper, lower] = this.readWord()
        const address = (upper << 8) + lower
        advancePR(this.PR, 1)
        switch(opcode) {
          case LD2:
            this.ld2(operands, address)
            break
          case ST:
            this.st(operands, address)
            break
          case LAD:
            this.lad(operands, address)
            break
          case ADDA2:
            this.adda2(operands, address)
            break
          case SUBA2:
            this.suba2(operands, address)
            break
          case ADDL2:
            this.addl2(operands, address)
            break
          case SUBL2:
            this.subl2(operands, address)
            break
          case AND2:
            this.and2(operands, address)
            break
          case OR2:
            this.or2(operands, address)
            break
          case XOR2:
            this.xor2(operands, address)
            break
          case CPA2:
            this.cpa2(operands, address)
            break
          case CPL2:
            this.cpl2(operands, address)
            break
          case SLA:
            this.sla(operands, address)
            break
          case SRA:
            this.sra(operands, address)
            break
          case SLL:
            this.sll(operands, address)
            break
          case SRL:
            this.srl(operands, address)
            break
          case JMI:
            this.jmi(operands, address)
            break
          case JNZ:
            this.jnz(operands, address)
            break
          case JZE:
            this.jze(operands, address)
            break
          case JUMP:
            this.jump(operands, address)
            break
          case JPL:
            this.jpl(operands, address)
            break
          case JOV:
            this.jov(operands, address)
            break
          case PUSH:
            this.push(operands, address)
            break
          case CALL:
            this.call(operands, address)
            break
          default:
            throw new Error(`unknown opcode. opcode=${opcode}, operands=${operands}`)
        }
    }

    const next = this.PR.lookup()
    if (next == -32678) {
      return false
    }
    return true
  }

  ld(operands: number) {
    const [n, m] = this.divide(operands)
    const v = this.gr(m)
    this.storeGr(n, v)
    this.FR.set(v)
  }

  ld2(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.memory.lookup(address + this.gr(x))
    this.storeGr(n, v)
    this.FR.set(v)
  }

  adda(operands: number) {
    const [n, m] = this.divide(operands)
    const v = this.gr(n) + this.gr(m)
    this.storeGr(n, v)
    this.FR.set(v)
  }

  adda2(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.gr(n) + this.memory.lookup(address + this.gr(x))
    this.storeGr(n, v)
    this.FR.set(v)
  }

  suba(operands: number) {
    const [n, m] = this.divide(operands)
    const v = this.gr(n) - this.gr(m)
    this.storeGr(n, v)
    this.FR.set(v)
  }

  suba2(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.gr(n) - this.memory.lookup(address + this.gr(x))
    this.storeGr(n, v)
    this.FR.set(v)
  }

  addl(operands: number) {
    const [n, m] = this.divide(operands)
    const v = this.grLogical(n) + this.grLogical(m)
    this.storeGrLogical(n, v)
    this.FR.setLogical(v)
  }

  addl2(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.grLogical(n) + this.memory.lookupLogical(address + this.gr(x))
    this.storeGrLogical(n, v)
    this.FR.setLogical(v)
  }

  subl(operands: number) {
    const [n, m] = this.divide(operands)
    const v = this.grLogical(n) - this.grLogical(m)
    this.storeGrLogical(n, v)
    this.FR.setLogical(v)
  }

  subl2(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.grLogical(n) - this.memory.lookupLogical(address + this.gr(x))
    this.storeGrLogical(n, v)
    this.FR.setLogical(v)
  }

  and(operands: number) {
    const [n, m] = this.divide(operands)
    const v = this.gr(n) & this.gr(m)
    this.storeGr(n, v)
    this.FR.set(v)
  }

  and2(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.gr(n) & this.memory.lookup(address + this.gr(x))
    this.storeGr(n, v)
    this.FR.set(v)
  }

  or(operands: number) {
    const [n, m] = this.divide(operands)
    const v = this.gr(n) | this.gr(m)
    this.storeGr(n, v)
    this.FR.set(v)
  }

  or2(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.gr(n) | this.memory.lookup(address + this.gr(x))
    this.storeGr(n, v)
    this.FR.set(v)
  }

  xor(operands: number) {
    const [n, m] = this.divide(operands)
    const v = this.gr(n) ^ this.gr(m)
    this.storeGr(n, v)
    this.FR.set(v)
  }

  xor2(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.gr(n) ^ this.memory.lookup(address + this.gr(x))
    this.storeGr(n, v)
    this.FR.set(v)
  }

  cpa(operands: number) {
    const [n, m] = this.divide(operands)
    const v = this.gr(n) - this.gr(m)
    this.FR.setByCPA(v)
  }

  cpa2(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.gr(n) - this.memory.lookup(address + this.gr(x))
    this.FR.setByCPA(v)
  }

  cpl(operands: number) {
    const [n, m] = this.divide(operands)
    const v = this.grLogical(n) - this.grLogical(m)
    this.FR.setLogicalByCPL(v)
  }

  cpl2(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.grLogical(n) - this.memory.lookupLogical(address + this.gr(x))
    this.FR.setLogicalByCPL(v)
  }

  pop(operands: number) {
    const [n, _] = this.divide(operands)
    const address = this.memory.lookupLogical(this.SP.lookupLogical())
    this.storeGrLogical(n, address)
    this.SP.storeLogical(this.SP.lookupLogical() + 1)
  }

  push(operands: number, address: number) {
    // address -> memory(SP-1)
    const [_, x] = this.divide(operands)
    this.SP.storeLogical(this.SP.lookupLogical() - 1)
    this.memory.storeLogical(this.SP.lookupLogical(), address + this.gr(x))
  }

  ret() {
    const sp = this.SP.lookupLogical()
    if (sp != END_ADDRESS) {
      const address = this.memory.lookupLogical(sp)
      this.PR.storeLogical(address)
      this.SP.storeLogical(sp+1)
    } else {
      this.PR.store(-32678)
    }
  }

  call(operands: number, address: number) {
    // PR -> memory(SP-1)
    // address -> PR
    const [_, x] = this.divide(operands)
    const wordLength = 2
    this.SP.storeLogical(this.SP.lookupLogical() - 1)
    this.memory.storeLogical(this.SP.lookupLogical(), this.PR.lookupLogical())
    this.PR.storeLogical(address + this.gr(x))
  }

  st(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = this.gr(n)
    this.memory.store(address + this.gr(x), v)
  }

  lad(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const v = address + this.gr(x)
    this.storeGr(n, v)
  }

  sla(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const isNegative = ((this.gr(n) >> 15) & 1) != 0
    const b = address + this.gr(x)
    let v = this.gr(n) << b
    if (isNegative) {
      v = v | (1 << 15)
    } else {
      v = v & ~(1 << 15)
    }
    let overflowFlag = false
    if (((this.grLogical(n) >> (16 -1 - b)) & 1) !== 0) {
      overflowFlag = true
    }
    this.storeGr(n, v)
    this.FR.setWithOverflowFlag(v, overflowFlag)
  }

  sra(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const b = address + this.gr(x)
    let v = this.gr(n) >> b
    let overflowFlag = false
    if (((this.grLogical(n) >> (b - 1)) & 1) !== 0) {
      overflowFlag = true
    }
    this.storeGr(n, v)
    this.FR.setWithOverflowFlag(v, overflowFlag)
  }

  sll(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const b = address + this.gr(x)
    let v = this.gr(n) << b
    let overflowFlag = false
    if (((this.grLogical(n) >> (16 - b)) & 1) !== 0) {
      overflowFlag = true
    }
    this.storeGr(n, v)
    this.FR.setWithOverflowFlag(v, overflowFlag)
  }

  srl(operands: number, address: number) {
    const [n, x] = this.divide(operands)
    const b = address + this.gr(x)
    let v = this.grLogical(n) >>> b
    let overflowFlag = false
    if (((this.grLogical(n) >> (b - 1)) & 1) !== 0) {
      overflowFlag = true
    }
    this.storeGr(n, v)
    this.FR.setWithOverflowFlag(v, overflowFlag)
  }

  jmi(operands: number, address: number) {
    const [_, x] = this.divide(operands)
    if (this.FR.sf()) {
      setPR(this.PR, address + this.gr(x))
    }
  }

  jnz(operands: number, address: number) {
    const [_, x] = this.divide(operands)
    if (!this.FR.zf()) {
      setPR(this.PR, address + this.gr(x))
    }
  }

  jze(operands: number, address: number) {
    const [_, x] = this.divide(operands)
    if (this.FR.zf()) {
      setPR(this.PR, address + this.gr(x))
    }
  }

  jump(operands: number, address: number) {
    const [_, x] = this.divide(operands)
    setPR(this.PR, address + this.gr(x))
  }

  jpl(operands: number, address: number) {
    const [_, x] = this.divide(operands)
    if (!this.FR.sf()) {
      setPR(this.PR, address + this.gr(x))
    }
  }

  jov(operands: number, address: number) {
    const [_, x] = this.divide(operands)
    if (this.FR.of()) {
      setPR(this.PR, address + this.gr(x))
    }
  }
}

File: wasm_interpreter/src/interpreter.cpp
#include <iostream>
#include "register.hpp"
#include "flag_register.hpp"
#include "memory.hpp"

// 1 word
const uint8_t NOP  = 0x00;
const uint8_t LD   = 0x14;
const uint8_t ADDA = 0x24;
const uint8_t SUBA = 0x25;
const uint8_t ADDL = 0x26;
const uint8_t SUBL = 0x27;
const uint8_t AND  = 0x34;
const uint8_t OR   = 0x35;
const uint8_t XOR  = 0x36;
const uint8_t CPA  = 0x44;
const uint8_t CPL  = 0x45;
const uint8_t POP  = 0x71;
const uint8_t RET  = 0x81;

// 2 words
const uint8_t LD2   = 0x10;
const uint8_t ST    = 0x11;
const uint8_t LAD   = 0x12;
const uint8_t ADDA2 = 0x20;
const uint8_t SUBA2 = 0x21;
const uint8_t ADDL2 = 0x22;
const uint8_t SUBL2 = 0x23;
const uint8_t AND2  = 0x30;
const uint8_t OR2   = 0x31;
const uint8_t XOR2  = 0x32;
const uint8_t CPA2  = 0x40;
const uint8_t CPL2  = 0x41;
const uint8_t SLA   = 0x50;
const uint8_t SRA   = 0x51;
const uint8_t SLL   = 0x52;
const uint8_t SRL   = 0x53;
const uint8_t JMI   = 0x61;
const uint8_t JNZ   = 0x62;
const uint8_t JZE   = 0x63;
const uint8_t JUMP  = 0x64;
const uint8_t JPL   = 0x65;
const uint8_t JOV   = 0x66;
const uint8_t PUSH  = 0x70;
const uint8_t CALL  = 0x80;

// no support
// const int SVC   = 0xF0;

const uint16_t END_ADDRESS = 0xE000;

class Interpreter
{
// private :
public :
  Register* gr0;
  Register* gr1;
  Register* gr2;
  Register* gr3;
  Register* gr4;
  Register* gr5;
  Register* gr6;
  Register* gr7;
  FlagRegister* fr;
  Register* pr;
  Register* sp;
  Memory* memory;

  Interpreter(
    Register* gr0,
    Register* gr1,
    Register* gr2,
    Register* gr3,
    Register* gr4,
    Register* gr5,
    Register* gr6,
    Register* gr7,
    FlagRegister* fr,
    Register* pr,
    Register* sp,
    Memory* memory
  ) {
    this->gr0 = gr0;
    this->gr1 = gr1;
    this->gr2 = gr2;
    this->gr3 = gr3;
    this->gr4 = gr4;
    this->gr5 = gr5;
    this->gr6 = gr6;
    this->gr7 = gr7;
    this->fr = fr;
    this->pr = pr;
    this->sp = sp;
    this->memory = memory;
  }

  void advancePR() {
    auto v = pr->lookupLogical();
    pr->storeLogical(v + 1);
  }

  std::pair<uint16_t, uint16_t> readWord() {
    try {
      auto currentPos = pr->lookupLogical();
      auto word = memory->lookupLogical(currentPos);
      uint8_t lowerByte = word & 0b11111111;
      uint8_t upperByte = word >> 8;
      return std::make_pair(upperByte, lowerByte);
    } catch (const std::exception& e) {
      std::cout << "Interpreter:: readWord() exception: " << e.what() << std::endl;
      return std::make_pair(0, 0);
    }
  }

  bool step() {
    if (pr->lookupLogical() == END_ADDRESS) {
      return false;
    }
    auto [opcode, operand] = readWord();
    advancePR();
    if (opcode == 0 && operand == 0) {
      return false;
    }

    switch (opcode) {
      case NOP:
        break;
      case LD:
        ld(operand);
        break;
      case ADDA:
        adda(operand);
        break;
      case SUBA:
        suba(operand);
        break;
      case ADDL:
        addl(operand);
        break;
      case SUBL:
        subl(operand);
        break;
      case AND:
        and1(operand);
        break;
      case OR:
        or1(operand);
        break;
      case XOR:
        xor1(operand);
        break;
      case CPA:
        cpa(operand);
        break;
      case CPL:
        cpl(operand);
        break;
      case POP:
        pop(operand);
        break;
      case RET:
        ret();
        break;
      default:
        auto [upper, lower] = readWord();
        advancePR();
        uint16_t address = (upper << 8) + lower;
        switch (opcode) {
          case LD2:
            ld2(operand, address);
            break;
          case ST:
            st(operand, address);
            break;
          case LAD:
            lad(operand, address);
            break;
          case ADDA2:
            adda2(operand, address);
            break;
          case SUBA2:
            suba2(operand, address);
            break;
          case ADDL2:
            addl2(operand, address);
            break;
          case SUBL2:
            subl2(operand, address);
            break;
          case AND2:
            and2(operand, address);
            break;
          case OR2:
            or2(operand, address);
            break;
          case XOR2:
            xor2(operand, address);
            break;
          case CPA2:
            cpa2(operand, address);
            break;
          case CPL2:
            cpl2(operand, address);
            break;
          case SLA:
            sla(operand, address);
            break;
          case SRA:
            sra(operand, address);
            break;
          case SLL:
            sll(operand, address);
            break;
          case SRL:
            srl(operand, address);
            break;
          case JMI:
            jmi(operand, address);
            break;
          case JNZ:
            jnz(operand, address);
            break;
          case JZE:
            jze(operand, address);
            break;
          case JUMP:
            jump(operand, address);
            break;
          case JPL:
            jpl(operand, address);
            break;
          case JOV:
            jov(operand, address);
            break;
          case PUSH:
            push(operand, address);
            break;
          case CALL:
            call(operand, address);
            break;
          default:
            throw std::runtime_error("Unknown opcode");
        }
        break;
    }
    auto next = pr->lookupLogical();
    if (next == END_ADDRESS) {
      return false;
    }
    return true;
  }

  std::pair<uint8_t, uint8_t> divide(uint8_t byte) {
    return std::make_pair(byte >> 4, byte & 0b1111);
  }

  Register* gr(uint8_t n) {
    switch (n) {
      case 0:
        return gr0;
      case 1:
        return gr1;
      case 2:
        return gr2;
      case 3:
        return gr3;
      case 4:
        return gr4;
      case 5:
        return gr5;
      case 6:
        return gr6;
      case 7:
        return gr7;
      default:
        throw std::runtime_error("Unknown register: GR" + std::to_string(n));
    }
  }

  int xaddr(int x) {
    if (1 <= x && x <= 7) {
      return gr(x)->lookupLogical();
    }
    return 0;
  }

  void ld(uint8_t operands) {
    auto [n, m] = divide(operands);
    auto v = gr(m)->lookup();
    gr(n)->store(v);
    fr->set(v);
  }

  void ld2(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = memory->lookup(address + xaddr(x));
    gr(n)->store(v);
    fr->set(v);
  }

  void adda(uint8_t operands) {
    auto [n, m] = divide(operands);
    auto v = gr(n)->lookup() + gr(m)->lookup();
    gr(n)->store(v);
    fr->set(v);
  }

  void adda2(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = gr(n)->lookup() + memory->lookup(address + xaddr(x));
    gr(n)->store(v);
    fr->set(v);
  }

  void suba(uint8_t operands) {
    auto [n, m] = divide(operands);
    auto v = gr(n)->lookup() - gr(m)->lookup();
    gr(n)->store(v);
    fr->set(v);
  }

  void suba2(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = gr(n)->lookup() - memory->lookup(address + xaddr(x));
    gr(n)->store(v);
    fr->set(v);
  }

  void addl(uint8_t operands) {
    auto [n, m] = divide(operands);
    auto v = gr(n)->lookupLogical() + gr(m)->lookupLogical();
    gr(n)->storeLogical(v);
    fr->set_logical(v);
  }

  void addl2(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = gr(n)->lookupLogical() + memory->lookupLogical(address + xaddr(x));
    gr(n)->storeLogical(v);
    fr->set_logical(v);
  }

  void subl(uint8_t operands) {
    auto [n, m] = divide(operands);
    auto v = gr(n)->lookupLogical() - gr(m)->lookupLogical();
    gr(n)->storeLogical(v);
    fr->set_logical(v);
  }

  void subl2(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = gr(n)->lookupLogical() - memory->lookupLogical(address + xaddr(x));
    gr(n)->storeLogical(v);
    fr->set_logical(v);
  }

  void and1(uint8_t operands) {
    auto [n, m] = divide(operands);
    auto v = gr(n)->lookup() & gr(m)->lookup();
    gr(n)->store(v);
    fr->set(v);
  }

  void and2(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = gr(n)->lookup() & memory->lookup(address + xaddr(x));
    gr(n)->store(v);
    fr->set(v);
  }

  void or1(uint8_t operands) {
    auto [n, m] = divide(operands);
    auto v = gr(n)->lookup() | gr(m)->lookup();
    gr(n)->store(v);
    fr->set(v);
  }

  void or2(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = gr(n)->lookup() | memory->lookup(address + xaddr(x));
    gr(n)->store(v);
    fr->set(v);
  }

  void xor1(uint8_t operands) {
    auto [n, m] = divide(operands);
    auto v = gr(n)->lookup() ^ gr(m)->lookup();
    gr(n)->store(v);
    fr->set(v);
  }

  void xor2(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = gr(n)->lookup() ^ memory->lookup(address + xaddr(x));
    gr(n)->store(v);
    fr->set(v);
  }

  void cpa(uint8_t operands) {
    auto [n, m] = divide(operands);
    auto v = gr(n)->lookup() - gr(m)->lookup();
    fr->set_by_cpa(v);
  }

  void cpa2(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = gr(n)->lookup() - memory->lookup(address + xaddr(x));
    fr->set_by_cpa(v);
  }

  void cpl(uint8_t operands) {
    auto [n, m] = divide(operands);
    auto v = gr(n)->lookupLogical() - gr(m)->lookupLogical();
    fr->set_logical_by_cpl(v);
  }

  void cpl2(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = gr(n)->lookupLogical() - memory->lookupLogical(address + xaddr(x));
    fr->set_logical_by_cpl(v);
  }

  void pop(uint8_t operands) {
    auto [n, x] = divide(operands);
    auto v = memory->lookup(sp->lookupLogical());
    gr(n)->storeLogical(v);
    sp->storeLogical(sp->lookupLogical() + 1);
  }

  void push(uint8_t operands, uint16_t address) {
    auto [_, x] = divide(operands);
    sp->storeLogical(sp->lookupLogical() - 1);
    memory->storeLogical(sp->lookupLogical(), address + xaddr(x));
  }

  void ret() {
    auto sp_value = sp->lookupLogical();
    if (sp_value != END_ADDRESS) {
      auto address = memory->lookupLogical(sp_value);
      pr->storeLogical(address);
      sp->storeLogical(sp_value + 1);
    } else {
      pr->store(END_ADDRESS);
    }
  }

  void call(uint8_t operands, uint16_t address) {
    auto [_, x] = divide(operands);
    sp->storeLogical(sp->lookupLogical() - 1);
    memory->storeLogical(sp->lookupLogical(), pr->lookupLogical());
    pr->storeLogical(address + xaddr(x));
  }

  void st(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto v = gr(n)->lookup();
    memory->store(address + xaddr(x), v);
  }

  void lad(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    gr(n)->store(address + xaddr(x));
  }

  void sla(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto isNegative = ((gr(n)->lookup() >> 15) & 1) != 0;
    auto b = address + xaddr(x);
    auto v = gr(n)->lookup() << b;
    if (isNegative) {
      v = v | (1 << 15);
    } else {
      v = v & ~(1 << 15);
    }
    auto overflow = false;
    if (((gr(n)->lookupLogical() >> (16 - 1 - b)) & 1) != 0) {
      overflow = true;
    }
    gr(n)->store(v);
    fr->set_with_overflow_flag(v, overflow);
  }

  void sra(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto b = address + xaddr(x);
    auto v = gr(n)->lookup() >> b;
    auto overflow = false;
    if (((gr(n)->lookupLogical() >> (b - 1)) & 1) != 0) {
      overflow = true;
    }
    gr(n)->store(v);
    fr->set_with_overflow_flag(v, overflow);
  }

  void sll(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto b = address + xaddr(x);
    auto v = gr(n)->lookup() << b;
    auto overflow = false;
    if (((gr(n)->lookup() >> (16 - b)) & 1) != 0) {
      overflow = true;
    }
    gr(n)->store(v);
    fr->set_with_overflow_flag(v, overflow);
  }

  void srl(uint8_t operands, uint16_t address) {
    auto [n, x] = divide(operands);
    auto b = address + xaddr(x);
    // auto v = gr(n)->lookupLogical() >>> b;
    auto v = gr(n)->lookupLogical() >> b;
    auto overflow = false;
    if (((gr(n)->lookup() >> (b - 1)) & 1) != 0) {
      overflow = true;
    }
    gr(n)->store(v);
    fr->set_with_overflow_flag(v, overflow);
  }

  void jmi(uint8_t operands, uint16_t address) {
    auto [_, x] = divide(operands);
    if (fr->sf()) {
      pr->store(address + xaddr(x));
    }
  }

  void jnz(uint8_t operands, uint16_t address) {
    auto [_, x] = divide(operands);
    if (!fr->zf()) {
      pr->store(address + xaddr(x));
    }
  }

  void jze(uint8_t operands, uint16_t address) {
    auto [_, x] = divide(operands);
    if (fr->zf()) {
      pr->store(address + xaddr(x));
    }
  }

  void jump(uint8_t operands, uint16_t address) {
    auto [_, x] = divide(operands);
    pr->store(address + xaddr(x));
  }

  void jpl(uint8_t operands, uint16_t address) {
    auto [_, x] = divide(operands);
    if (!fr->sf()) {
      pr->store(address + xaddr(x));
    }
  }

  void jov(uint8_t operands, uint16_t address) {
    auto [_, x] = divide(operands);
    if (fr->of()) {
      pr->store(address + xaddr(x));
    }
  }
};
